### ISSUE TITLE
Update SDK 23 release notes to indicate how to migrate ExpoKit iOS Launch xib

### ISSUE THREAD

**Issue by asgeo1** [expo/expo#985]
I have a detached Expo app, and have been trying to update it to the latest version of Expo.

Expo 20.0.0 => 23.0.0
Expo Kit 1.19.0 => 2.1.1
React 16.0.0-alpha.12 => 16.0.0
React Native to 0.50

I've followed all of the instructions, and my app compiles OK.  But when I run it, it crashes with `libc++abi.dylib: terminating with uncaught exception of type NSException`, with no other information.

```
2017-11-21 18:28:30:104 mau[7403:5275900] Expo Remote Notification services won't work in an ExpoKit app because Expo can not manage your APNS certificates.
2017-11-21 18:28:30:153 mau[7403:5275900] 2017-11-21 18:28:30.153 [warn][tid:main][RCTBridge.m:121] Class RCTDevMenu was not exported. Did you forget to use RCT_EXPORT_MODULE()?
2017-11-21 18:28:30:154 mau[7403:5275900] 2017-11-21 18:28:30.154 [warn][tid:main][RCTBridge.m:121] Class EXFrameExceptionsManager was not exported. Did you forget to use RCT_EXPORT_MODULE()?
2017-11-21 18:28:30:176 mau[7403:5275900] 2017-11-21 18:28:30.177 [warn][tid:main][RCTBridge.m:121] Class EXTest was not exported. Did you forget to use RCT_EXPORT_MODULE()?
2017-11-21 18:28:30:200 mau[7403:5275900] 2017-11-21 18:28:30.200 [warn][tid:main][RCTModuleData.mm:69] Module EXAppState requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:200 mau[7403:5275900] 2017-11-21 18:28:30.201 [warn][tid:main][RCTModuleData.mm:69] Module EXDeviceMotion requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:200 mau[7403:5275900] 2017-11-21 18:28:30.201 [warn][tid:main][RCTModuleData.mm:69] Module EXConstants requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:201 mau[7403:5275900] 2017-11-21 18:28:30.201 [warn][tid:main][RCTModuleData.mm:69] Module EXFileSystem requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:201 mau[7403:5275900] 2017-11-21 18:28:30.202 [warn][tid:main][RCTModuleData.mm:69] Module EXSecureStore requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:201 mau[7403:5275900] 2017-11-21 18:28:30.202 [warn][tid:main][RCTModuleData.mm:69] Module EXKernelModule requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:202 mau[7403:5275900] 2017-11-21 18:28:30.203 [warn][tid:main][RCTModuleData.mm:69] Module EXAdSettingsManager requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:203 mau[7403:5275900] 2017-11-21 18:28:30.203 [warn][tid:main][RCTModuleData.mm:69] Module EXAnimationViewManager requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:203 mau[7403:5275900] 2017-11-21 18:28:30.203 [warn][tid:main][RCTModuleData.mm:69] Module EXAppLoadingManager requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:203 mau[7403:5275900] 2017-11-21 18:28:30.203 [warn][tid:main][RCTModuleData.mm:69] Module EXAV requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:204 mau[7403:5275900] 2017-11-21 18:28:30.204 [warn][tid:main][RCTModuleData.mm:69] Module EXCameraManager requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:204 mau[7403:5275900] 2017-11-21 18:28:30.204 [warn][tid:main][RCTModuleData.mm:69] Module EXGLObjectManager requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:204 mau[7403:5275900] 2017-11-21 18:28:30.205 [warn][tid:main][RCTModuleData.mm:69] Module EXGLViewManager requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:204 mau[7403:5275900] 2017-11-21 18:28:30.205 [warn][tid:main][RCTModuleData.mm:69] Module EXImagePicker requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:205 mau[7403:5275900] 2017-11-21 18:28:30.205 [warn][tid:main][RCTModuleData.mm:69] Module EXLocation requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:205 mau[7403:5275900] 2017-11-21 18:28:30.206 [warn][tid:main][RCTModuleData.mm:69] Module EXNativeAdManager requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:206 mau[7403:5275900] 2017-11-21 18:28:30.206 [warn][tid:main][RCTModuleData.mm:69] Module EXPermissions requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:206 mau[7403:5275900] 2017-11-21 18:28:30.206 [warn][tid:main][RCTModuleData.mm:69] Module EXVideoManager requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:207 mau[7403:5275900] 2017-11-21 18:28:30.207 [warn][tid:main][RCTModuleData.mm:69] Module RNBranch requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:207 mau[7403:5275900] 2017-11-21 18:28:30.207 [warn][tid:main][RCTModuleData.mm:69] Module RNBranchEventEmitter requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:208 mau[7403:5275900] 2017-11-21 18:28:30.208 [warn][tid:main][RCTModuleData.mm:69] Module RNGestureHandlerModule requires main queue setup since it overrides `constantsToExport` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
2017-11-21 18:28:30:210 mau[7403:5275900] 2017-11-21 18:28:30.210 [warn][tid:main][RCTModuleData.mm:69] Module RCTTVNavigationEventEmitter requires main queue setup since it overrides `init` but doesn't implement `requiresMainQueueSetup. In a future release React Native will default to initializing all native modules on a background thread unless explicitly opted-out of.
```

Using breakpoints, I traced it to this line of code.  After this line, the app will crash:

![image](https://user-images.githubusercontent.com/702212/33059788-e27a412e-cee9-11e7-8c8c-d9e62f868e79.png)

After this line, the app will crash with `libc++abi.dylib: terminating with uncaught exception of type NSException`

Some other people were reporting this exact same issue here: https://github.com/facebook/react-native/issues/16447  so I'm not the only one.

Any ideas?

**Comment by Ranatchai** [comment]
+1 I have the same issue.

**Comment by terribleben** [comment]
Hi, that error message is just a generic description printed by Xcode for every possible exception. It means "there was an exception." So it doesn't tell us much about why the app crashed. 

If possible, can you provide the stack trace for the exception?

Can you also verify that you followed the SDK-specific upgrade instructions for ExpoKit? e.g. moving from pre-SDK-22 to 22, the yoga dependency changed, as detailed [at the bottom of this post](https://blog.expo.io/expo-sdk-v22-0-0-is-now-available-7745bfe97fc6). If you are coming from a much older version then you will probably also need to add the `BatchedBridge` dependency.

**Comment by cellis** [comment]
I get this when I run this code ( by pressing a button )

```javascript
const { type, token } = await Expo.Facebook.logInWithReadPermissionsAsync('***app_id_here***', {
      permissions: ['public_profile'],
    });
```

**Comment by asgeo1** [comment]
I suspect that @cellis 's issue is not the same as mine. Different Expo SDK / ExpoKit versions, and difference in how it is cashing (mine is during startup).

As I say, some people on https://github.com/facebook/react-native/issues/16447 where having the exact same issue as I'm describing now though, crashing during startup after doing the same upgrade steps.

Sadly there is no stack trace for my issue. It's an unhandled exception and is getting caught in main.m, so there is no indication where the crash originally occurred.

![image](https://user-images.githubusercontent.com/702212/33099629-99e4e42e-cf65-11e7-9412-995be7dee776.png)


That's why I tried setting various breakpoints to see how far the program could run before it would actually crash.   My original screenshot above shows that.

So line EXViewController.m:126 is the line which it will actually crash at

![image](https://user-images.githubusercontent.com/702212/33099773-232b0088-cf66-11e7-8050-e26fba666509.png)

I was hoping that might provide a clue to why it's crashing?


Yes, I followed all of the instructions for upgrading from SDK 20 to SDK 23, including the yogo change.

**Comment by terribleben** [comment]
Can you go to Product -> Scheme -> Edit Scheme... -> Run -> Environment Variables, and uncheck the `OS_ACTIVITY_MODE: disable` box if you see it checked, then try again? Then see if your lower-right-hand Xcode console prints a stack trace other than "terminating with uncaught exception"

**Comment by asgeo1** [comment]
@terribleben - yes, that works for showing the actual error!

```
2017-11-22 09:41:06.543079+1100 mau[7504:5331077] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Could not load NIB in bundle: 'NSBundle </var/containers/Bundle/Application/184D9E1E-563E-4138-9049-C926308F6D18/mau.app> (loaded)' with name 'LaunchScreen''
*** First throw call stack:
(0x18454bd38 0x183a60528 0x18454bc80 0x18dee15e4 0x18dee964c 0x100b5b3e0 0x18d95bbfc 0x18d95b7d4 0x100b5bf54 0x100b3aecc 0x100b5bd24 0x1009a08e4 0x18d9cb050 0x18dbbe898 0x18dbc36e4 0x18de51454 0x18e1211f0 0x18de510b8 0x18de51928 0x18e5ba6e8 0x18e5ba58c 0x18e3369c0 0x18e4cbfc8 0x18e336870 0x18e120850 0x18dbc1e28 0x18dfc56ec 0x186bed768 0x186bf6070 0x102dad45c 0x102db9b74 0x186c21a04 0x186c216a8 0x186c21c44 0x1844f4358 0x1844f42d8 0x1844f3b60 0x1844f1738 0x1844122d8 0x1862a3f84 0x18d9bf880 0x1009a05dc 0x183f3656c)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

Strange... there is a file called `LaunchScreen.storyboard`

![image](https://user-images.githubusercontent.com/702212/33100827-ed723f0c-cf69-11e7-8c34-3928ee2fba32.png)

**Comment by terribleben** [comment]
Ah, thanks for the report. We did change how the Launch image works on iOS between SDK 22 and SDK 23. Since we can't re-detach for you, your project still contains a reference to the old `.storyboard` file instead of the new `.xib` file. I recommend [downloading this xib file](https://github.com/expo/expo/blob/master/exponent-view-template/ios/exponent-view-template/Supporting/LaunchScreen.xib) and adding it to your project's Supporting directory, then removing any old references to the `.storyboard` file.

![image](https://user-images.githubusercontent.com/1316332/33101112-d12a1412-cecb-11e7-943c-fa7d511519ed.png)

And I'll go update our SDK 23 release notes to include these instructions. I can also try to write a fix which at least prevents a crash when somebody has not updated this yet.

**Comment by asgeo1** [comment]
Thanks Ben!

Yes, that means that I also need to add `launch_background_image.png` and `launch_icon.png`...

I get a different error now:

`Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[OS_dispatch_data data]: unrecognized selector sent to instance 0x13687da00`

![image](https://user-images.githubusercontent.com/702212/33101866-3a786bce-cf6e-11e7-9312-f994f8d782c2.png)

Any ideas?

**Comment by asgeo1** [comment]
Well, I've just tried updating an brand new empty (and detatched) Expo project from SDK 22 to SDK 23.

All I had to do was address the launch image / zib items as you mentioned, and then it worked.

So whatever my latest issue is, is probably something else, and possibly not even Expo related.  Hard to tell...

It might just be easier for me to copy all my app code into the newly created empty project.

**Comment by terribleben** [comment]
Ok. I'm not familiar with that last thing you pasted about `OS_dispatch_data`, but a quick google search indicates it might be something about the socket library you're using, and maybe transient.

I've updated the [release notes](https://blog.expo.io/expo-sdk-v23-0-0-is-now-available-be0a8c655414) to include instructions for updating the .xib file, so I'll close this issue for now. Let me know if you're still having trouble.
